mountpoint-s3/src/prefetch/checksummed_bytes.rs

```diff
@@ -0,0 +1,250 @@
+        }
+    }
+
+    /// Convert the `ChecksummedBytes` into `Bytes`, data integrity will be validated before converting.
+    ///
+    /// Return `IntegrityError` on data corruption.
+    pub fn into_bytes(self) -> Result<Bytes, IntegrityError> {
+        #[cfg(feature = "checksum")]
+        self.validate()?;
+
+        Ok(self.curr_slice)
+    }
+
+    /// Returns the number of bytes contained in this `ChecksummedBytes`.
+    pub fn len(&self) -> usize {
+        self.curr_slice.len()
+    }
+
+    /// Returns true if the `ChecksummedBytes` has a length of 0.
+    pub fn is_empty(&self) -> bool {
+        self.curr_slice.is_empty()
+    }
+
+    /// Split off the checksummed bytes at the given index.
+    ///
+    /// Afterwards self contains elements [0, at), and the returned Bytes contains elements [at, len).
+    ///
+    /// This operation just increases the reference count and sets a few indices,
+    /// so there will be no validation and the checksum will not be recomputed.
+    pub fn split_off(&mut self, at: usize) -> ChecksummedBytes {
+        let new_bytes = self.curr_slice.split_off(at);
+        Self {
+            orig_bytes: self.orig_bytes.clone(),
+            curr_slice: new_bytes,
+            checksum: self.checksum,
+        }
+    }
+
+    /// Append the given checksummed bytes to current `ChecksummedBytes`, ensure that data integrity will
+    /// be validated.
+    ///
+    /// Return `IntegrityError` on data corruption.
+    pub fn extend(&mut self, extend: ChecksummedBytes) -> Result<(), IntegrityError> {
+        let curr_len = self.curr_slice.len();
+        let total_len = curr_len + extend.len();
+
+        let mut bytes_mut = BytesMut::with_capacity(total_len);
+        bytes_mut.extend_from_slice(&self.curr_slice);
+        bytes_mut.extend_from_slice(&extend.curr_slice);
+        let new_bytes = bytes_mut.freeze();
+        let new_checksum = crc32c::checksum(&new_bytes);
+        let new_checksummed_bytes = ChecksummedBytes::new(new_bytes, new_checksum);
+
+        // Validate data integrity with checksum bracketing.
+        #[cfg(feature = "checksum")]
+        {
+            // 1. repeat the operation, which means copying into a new buffer in this case.
+            let mut bytes_mut_dup = BytesMut::with_capacity(total_len);
+            bytes_mut_dup.extend_from_slice(&self.curr_slice);
+            bytes_mut_dup.extend_from_slice(&extend.curr_slice);
+            let new_bytes_dup = bytes_mut_dup.freeze();
+            let new_checksum_dup = crc32c::checksum(&new_bytes_dup);
+
+            // 2. compare the checksum between the two transformations.
+            if new_checksum != new_checksum_dup {
+                return Err(IntegrityError::ChecksumMismatch(new_checksum, new_checksum_dup));
+            }
+
+            // 3. validate original buffers to make sure that the data we have copied are still valid.
+            self.validate()?;
+            extend.validate()?;
+        }
```

Comment:
The "do it twice" thing is required for more general transformations, but copying with CRC is a special case that shouldn't require it, because the checksums of the inputs and outputs are related. This code should be able to just do this:
1. Create C = A ++ B by copying from A and B
2. Compute CRC32(C) from C
3. Compute CRC32(A ++ B) = CRC32(A) # CRC32(B), where # is the combination function for CRC. We can do this cheaply since A and B are both ChecksummedBytes.
4. Fail if CRC32(C) != CRC32(A ++ B).

If A was corrupted into A' before copying into C, we'll find out when we compute CRC32(C) = CRC32(A' ++ B) and find that it doesn't match CRC32(A ++ B) = CRC32(A)  # CRC32(B). Same for B/B'. If C was corrupted after copying, it's the same thing -- CRC32(C') won't match CRC32(A)  # CRC32(B).

Notice this only works because of step (3), where we can compute the expected checksum of the output from the expected checksums of the inputs. For most transformations, and most checksum algorithms, that isn't possible. But this is a special case.

Reply:
I think that's true only when we copy the whole buffer. The problem here is that A and B could be slices of bigger buffers. For example, if A is a slice of X and B is a slice of Y then we can't compute CRC32(A ++ B) because we only have checksums of X and Y.

We can optimize it a bit here though, by checking that if A and B are original buffers we can skip the "do it twice" part.

Reply:
Hmm. I wonder how much we care about `split_off` performance versus making this operation faster, then. I guess we're happy because `extend` is likely to be the corner case (the slow path for `read`)?